### PR TITLE
Add statistics panel

### LINKED ISSUE #47: Add statistics panel
Adding a small statistics panel that lists the number of different destinations, number of countries, etc. that can slide in and out on the right side of the screen could be cool.

### PR DESCRIPTION
**Pull request by AndrewLester**
Closes #47 
Data is processed using a Google Apps Script on every form submission.
Here is the closed panel:
Screen Shot 2021-04-23 at 12 24 11 PM
Here is the open panel (black background):
Screen Shot 2021-04-23 at 12 23 56 PM
Here is the open panel (gray background):
Screen Shot 2021-04-23 at 12 24 05 PM

Thoughts?

### CODE REVIEW

**Comment by ErikBoesen** [comment]
This is dope! Some thoughts:
* Is it actually too resource intensive to compute these stats in JS rather than in a Google Apps script? I would prefer tracking code within the GH repository to the maximum extent possible
* The arrow on the right doesn't draw the eye. Perhaps we could move the trigger to be around the options panel in the top left, and maybe have a grey background around the button so it's easier to see against complex backgrounds?

File: img/chevron_left_white_24dp.svg [hunk 1]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 0 24 24" width="24px" fill="#FFFFFF"><path d="M0 0h24v24H0V0z" fill="none"/><path d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z"/></svg>
```
Comment (ErikBoesen) [inline img/chevron_left_white_24dp.svg]: Shorten this filename. `left_arrow.svg`? `arrow.svg`?

File: index.html [hunk 2]
```diff
@@ -42,6 +42,19 @@
             <p class="branding-info">Senior project by <a href="https://erikboesen.com">Erik Boesen</a> '19</p>
             <p class="branding-info">Maintained by GMHS <a target="_blank" rel="noopener" href="https://apc-gm.com">A.P.C.</a></p>
         </div>
+        <div id="statistics">
+            <img id="statistics-button" src="/img/chevron_left_white_24dp.svg">
+            <h3 id="statistics-header">
+                Class of <span id="statistics-header-class" class="data"></span> stats
+            </h3>
+            <ul id="statistics-stats">
+                <li><span id="stats-students" class="data"></span> students displayed</li>
```
Comment (ErikBoesen) [inline index.html]: Nit: `data` is a fairly vague name. Perhaps choose something more specific (though still succinct)?

File: scripts.js [hunk 3]
```diff
@@ -6,6 +6,19 @@ var options = {
     precedence: document.getElementById('option-precedence'),
 };
 
+var statistics = {
+    statistics: document.getElementById('statistics'),
```
Comment (ErikBoesen) [inline scripts.js]: Maybe `container`? This is fine though if you can't think of anything else.

File: scripts.js [hunk 4]
```diff
@@ -79,6 +93,11 @@ async function fetchDataDocuments() {
 
         options.year.prepend(option);
     }
+
+    for (let statsObj of tabletop.sheets('stats').all()) {
```
Comment (ErikBoesen) [inline scripts.js]: I instinctively recoil at variable names that include their type (`Obj`). Perhaps think of a different name?

File: scripts.js [hunk 5]
```diff
@@ -79,6 +93,11 @@ async function fetchDataDocuments() {
...
+        stats.set(statsObj['Year'], statsObj);
+        delete statsObj['Year'];
```
Comment (ErikBoesen) [inline scripts.js]: Why is this necessary?

File: scripts.js [hunk 6]
```diff
@@ -133,7 +153,8 @@ function displayMap(year, firstLoad) {
             sleep(firstLoad ? 0 : 300)
         ])
             .then(() => buildInstitutionData(year))
-            .then((institutions) => placeMarkers(institutions)),
+            .then((institutions) => placeMarkers(institutions))
```
Comment (ErikBoesen) [inline scripts.js:145]: Optional: drop paren around `(institutions)`

File: scripts.js [hunk 7]
```diff
@@ -298,10 +319,28 @@ options.pan.addEventListener('click', function (event) {
     panToMarkers = event.target.checked;
 });
 
-options.precedence.addEventListener('change', (event) => {
+options.precedence.addEventListener('change', function (event) {
```
Comment (ErikBoesen) [inline scripts.js]: Elsewhere we have no space between `function(`

File: scripts.js [hunk 8]
```diff
@@ -6,6 +6,19 @@ var options = {
     precedence: document.getElementById('option-precedence'),
 };
 
+var statistics = {
```
Comment (ErikBoesen) [inline scripts.js]: Since we now have a lot of elements declared, could we put `mapElement`, `options`, and `statistics` under a single object called `elements`/`elem`/`e`/`page` or something to that effect? I think it could make the code a little cleaner.

File: scripts.js [hunk 9]
```diff
@@ -298,10 +319,28 @@ options.pan.addEventListener('click', function (event) {
...
+    statistics.stats.students.textContent = students.get(year).length;
+
+    if (!stats.get(year)['Number of destinations']) {
+        statistics.statistics.classList.add('no-data');
+    }
+
+    statistics.stats.destinations.textContent = stats.get(year)['Number of destinations'];
+    statistics.stats.states.textContent = stats.get(year)['Number of states'];
+    statistics.stats.countries.textContent = stats.get(year)['Number of countries'];
+}
```
Comment (ErikBoesen) [inline scripts.js]: We are adding a lot of new code here, and `scripts.js` has been getting bigger for a while now. Do you think this new statistics functionality could better be placed in a new file? (And more generally, should we consider splitting existing code from `scripts.js` into separate files?

File: style.css [hunk 10]
```diff
@@ -159,6 +159,51 @@ div a[rel="noopener"] img {
...
+#statistics-header {
+    margin-bottom: 15px;
+}
+#statistics-stats {
+    list-style: none;
+    line-height: 2;
+}
+#statistics span.data {
+    font-weight: bold;
+}
```
Comment (ErikBoesen) [inline style.css]: Same thing here. Perhaps we should spin this off into a `statistics.css` file?

**Review by ErikBoesen — CHANGES_REQUESTED** [review]
Mostly just naming/styling nitpicks. Good job!

File: scripts.js (hunk 5 above)
Comment (ErikBoesen) [reply scripts.js]: Normally JavaScript should handle this kind of operation automatically, is there a specific memory concern here?

File: scripts.js (hunk 8 above)
Comment (ErikBoesen) [reply scripts.js]: If we split `scripts.js` into multiple files, this might not be a good choice. Although we could still do something like
```js
elements.statistics = {
    // ...
};
```
in a hypothetical `statistics` file.

File: scripts.js (hunk 4 above)
Comment (AndrewLester) [reply scripts.js:97]: > * Is it actually too resource intensive to compute these stats in JS rather than in a Google Apps script? I would prefer tracking code within the GH repository to the maximum extent possible
It's not the resource intensity that is the problem, but the computation to find the number of states/countries involves a geocode that needs to be run upon every submission to the forms. The Google Apps Script has a onFormSubmit trigger that serves this purpose nicely. I can add the code for it in this repo, but it wouldn't really do much and it'd be a bit of a duplication. I tried to put it in the Google Drive but for some reason Google Apps Scripts don't appear...

> * The arrow on the right doesn't draw the eye. Perhaps we could move the trigger to be around the options panel in the top left, and maybe have a grey background around the button so it's easier to see against complex backgrounds?

I'll give it a try on both the left and the right, but it certainly does need a gray background.

**Comment by AndrewLester** [comment]
> * Is it actually too resource intensive to compute these stats in JS rather than in a Google Apps script? I would prefer tracking code within the GH repository to the maximum extent possible

It's not the resource intensity that is the problem, but the computation to find the number of states/countries involves a geocode that needs to be run upon every submission to the forms. The Google Apps Script has an onFormSubmit trigger that serves this purpose nicely. I can add the code for it in this repo, but it wouldn't really do much and it'd be a bit of a duplication. I tried to put it in the Google Drive but for some reason Google Apps Scripts don't appear...

I think the fact that this code is external from the repo is *alright* at least, since the scripts.js file just reads the processed data from a spreadsheet. Essentially, the fact that the Google Apps Script does any processing is unknown to the scripts.js file, and someone could just manually put it the stats for all it cares.

> * The arrow on the right doesn't draw the eye. Perhaps we could move the trigger to be around the options panel in the top left, and maybe have a grey background around the button so it's easier to see against complex backgrounds?

I'll give it a try on both the left and the right, but it certainly does need a gray background.

File: scripts.js (hunk 5 above)
Comment (AndrewLester) [reply scripts.js]: Edit: The delete here is just to clean up the `statsObj` variable which contains all the statistics for a certain year. By nature of the spreadsheet, it also contains a key called "Year," which refers to the year, but I'd like this to be the key to the `stats` map rather than data that is available to the stats object stored in the map. No memory issue, I just thought it would be nicer to remove it from the object.

File: scripts.js (hunk 9 above)
Comment (AndrewLester) [reply scripts.js]: I like this idea, but I don't usually like including several JS files in the same HTML file and sharing functions between them (when we almost inevitably will need to). I think I'll do it for this stats stuff, but in the future maybe some kind of module system would work. I'll have to look into it.

File: scripts.js (hunk 6 above)
Comment (AndrewLester) [reply scripts.js:145]: I prefer this in this case.

Comment (ErikBoesen) [reply scripts.js:145]: That's legit

File: scripts.js (hunk 9 above)
Comment (ErikBoesen) [reply scripts.js]: `util.js`?

**Comment by ErikBoesen** [comment]
I would prefer that we at least copy the apps script code into this repo in case we were to lose it somehow.

**Comment by JonathanOppenheimer** [comment]
I think Erik already covered the actual coding stuff which I'm far less qualified to give advice on anyway but as this page gets more crowded, I had a couple of design thoughts: 

To start, I'd think the order of importance for top layer items goes (and if you guys disagree the same sorta thoughts apply regardless): 

year selection > destination submission > stats tab > full screen > zoom

When you look at a page, most western readers typically go top left to top right, to the bottom left, to bottom right
Screen Shot 2021-04-23 at 6 18 27 PM
(as you can see I am excellent at design) 

The top left and bottom right areas normally have the most important items, and you'll see this everywhere -- like a GitHub comment box! 
Screen Shot 2021-04-23 at 6 06 24 PM

I think the current layout with all of the items discussed above could definitely be improved by being rearranged. This would put emphasis on the more important stuff, catching the eye more immediately. It'd also help make the arrow more visible.

Low-quality paint job below -- if this doesn't make any sense, please disregard haha. 
115901719-8cf1b100-a42f-11eb-83ae-4c4f9272d140
(I think it's more easily understood if you take andrew's screenshot, and mine and then flip back and forth -- I think the edit is easier on the eyes)

**Review by JonathanOppenheimer — DISMISSED** [review]
(feedback above)

**Comment by AndrewLester** [comment]
I like the layout change, but I think leaving Google's buttons where they are is important for user recognition. So, I left the stats panel where it is on desktop since I already changed the css to make it look like a Google button.

However, on mobile I think we need more space for people to see the map, so I did choose to move the panel below the options panel as you showed.

**Comment by AndrewLester** [comment]
Ok turns out it looks better that way on desktop too. Final positioning is below options panel on all devices.

**Comment by JonathanOppenheimer** [comment]
> I like the layout change, but I think leaving Google's buttons where they are is important for user recognition. So, I left the stats panel where it is on desktop since I already changed the css to make it look like a Google button.

Yeah if the Google menu items take deference that's totally fair

**Review by JonathanOppenheimer — APPROVED** [review]

**Comment by ErikBoesen** [comment]
I don't think it's very feasible to move Google's buttons, but otherwise I like that take.

File: css/style.css [hunk 11]
```diff
@@ -70,6 +70,7 @@ div a[rel="noopener"] img {
     width: 200px;
     height: auto;
     background: #222;
+    z-index: 5;
```
Comment (ErikBoesen) [inline css/style.css:73]: This is a very civilized z-index

File: js/scripts.js [hunk 12]
```diff
@@ -77,7 +80,12 @@ async function fetchDataDocuments() {
             option.selected = true;
         }
 
-        options.year.prepend(option);
+        elements.options.year.prepend(option);
+    }
+
+    for (let yearStats of tabletop.sheets('stats').all()) {
+        stats.set(yearStats['Year'], yearStats);
+        delete yearStats['Year'];
```
Comment (ErikBoesen) [inline js/scripts.js:88]: Still not clear on why this is necessary.

**Review by ErikBoesen — APPROVED** [review]

**Review by RohanVSuri — APPROVED** [review]
Looks very good Andrew!

File: css/style.css (hunk 11 above)
Comment (AndrewLester) [reply css/style.css:73]: When only one element needs to above all else then I use 999999, when I need to style two elements I turn to reasonable numbers.

File: js/scripts.js (hunk 12 above)
Comment (AndrewLester) [reply js/scripts.js:88]: It's not really necessary, just makes sure the `yearStats` object only includes actual stats (states, countries, etc.) and not the year of those stats, since the year is already the key of the `Map` the stats are in.
